### ISSUE TITLE
Proposal: `usize` definition should be refined

### ISSUE THREAD

**Issue by MasterQ32** [ziglang/zig#5185]
I came across the definition of `usize`, which is currently defined as *unsigned pointer sized integer* and a question arose: Size of *what* pointer? Function pointer? Pointer to constant data? Pointer to mutable data?

For most platforms, the answer is simple: *There is only one address space*.

But as Zig tries to target *all* platforms, we should bear in mind that this is not true for all platforms. 

**Case Study:**
Zig supports AVR at the moment which has two memory spaces:

- Data
- Code

Both memory spaces have different adressing modes which can be used with the `Z` register, which is a 16 bit register. Thus, we could concloud that the pointer size is 16 bit. But the AVR instruction set also has a `RAMPZ` register that is prepended to the `Z` register to extend the memory space to 24 bit. Some modern AVRs have more than 128k ROM (e.g. Mega2560). This means that the effective pointer size 24 bit.

The same problem arises when targeting the 8086 CPU with segmentation. The actual pointer is a 20 bit value that is calculated by combining two 16 bit values (segment + offset).

**Problem:**
`usize` communicates that it stores the *size* of something, not the *address*. Right now, usize can contain values larger than the biggest (continously) adressable object in the language and it takes up more space than needed.

C has two distinct types for that reason:
- `size_t` (can store the size of an adressable object)
- `uintptr_t` (can store *any* pointer)

AVR-GCC solves the problem of 24 bit pointers by ignoring it and creates shims for functions that are linked beyond the 128k boundary. Data beyond the 64k boundary cannot be adressed and afaik LLVM has the same restriction. I don't think Zig should *ignore* such platform specifics and should be able to represent them correctly.

**Proposal:**
Redefine `usize` to be *can store the size of any object or array* and introduce a new type `upointer` that is *pointer sized integer*. Same for `isize` and `ipointer`.

It should also be discussed if a `upointer` will have a guaranteed unique representation or may be ambiguous ("storing a linear address or storing segment + descriptor")?

**Note:**
I'm not quite sure about all of this yet as this is a very special case that only affects *some* platforms whereas most platforms don't have the *object size is not pointer size* restriction.

**Resources:**
- [AVR Instruction Set](http://ww1.microchip.com/downloads/en/devicedoc/atmel-0856-avr-instruction-set-manual.pdf)
- [Accessing Memory Outside of the 64K Range](https://imagecraft.com/help/ICCV8AVR/iccavr/6-programmingavr/accessing_memory_outside_of_the_64k_range.htm)
- [8086 Adressing](https://wiki.osdev.org/Real_Mode#Memory_Addressing)
- [Explanation/Discussion on SO](https://stackoverflow.com/a/1464194)

**Comment by LemonBoy** [comment]
Counter proposal:
> Size of what pointer? Function pointer? Pointer to constant data? Pointer to mutable data?

The maximum across all the address spaces. This way we can also keep the ptr-to-usize (and usize-to-ptr) relationship (with the help of the `addresspace` pointer metadata).

**Comment by MasterQ32** [comment]
> The maximum across all the address spaces. This way we can also keep the ptr-to-usize (and usize-to-ptr) relationship (with the help of the addresspace pointer metadata).

I thought about that, and it has one problem: It will waste a lof of space. `@ptrToInt` and `@intToPtr` should use `upointer`, but if i want to store a *size* of something (which is the standard case), i should use `usize`.

This has two advantages: `upointer` should always store pointers, `usize` should always store object size. So `@sizeOf` will return `usize`.

Otherwise i would waste 50% of my memory with zero padded bytes by storing pointers where i could've used a type only having half of the size.

**Comment by MasterQ32** [comment]
Added some changes and updates to the original proposal

**Comment by JesseRMeyer** [comment]
Since Zig supports arbitrarily sized integer types, each OS could define the bit length of their virtual memory system.  This would reduce the well known 'pointer bloat' in the executable.

**Comment by ityonemo** [comment]
would you have to change the default typing rules on certain arithmetic events?  like would the following operations make sense?

usize + usize OK
upointer + upointer NO
upointer + usize OK
usize * iX OK
usize * usize NO
usize * upointer NO
upointer * iX NO

**Comment by MasterQ32** [comment]
> would you have to change the default typing rules on certain arithmetic events? like would the following operations make sense?

No. I don't think this is something that is such a huge error source that it would be a benefit more than a hassle. Subtracting (and thus) adding values of `upointer` is quite helpful some times, also note that `upointer - upointer` may still not fit into a `usize`.

**Comment by floopfloopfloopfloopfloop** [comment]
Is there any real use-case for `isize` and `ipointer`? The only time I can imagine a negative value would be used in those contexts is to check for overflow, or something esoteric. The former can be done with existing operators. The latter would need a Real World Usecase.

The purpose we're trying to imply is "size of a pointer" and "size of a datablock", so I would simply lean toward `psize` and `msize` storing "pointer sized unsigned [~~integer~~](https://github.com/ziglang/zig/issues/1974#issuecomment-620336282)" and "maximum-allocatable-memory sized unsigned [~~integer~~](https://github.com/ziglang/zig/issues/1974#issuecomment-620336282)".

**Comment by kyle-github** [comment]
There are OSes that treat pointers as signed.   Solaris maybe?   That's not a big market :-)

**Comment by floopfloopfloopfloopfloop** [comment]
Real Life Use Case? :thinking:

**Comment by floopfloopfloopfloopfloop** [comment]
Disclaimer: I don't think the following is a well thought out idea as it stands, but my brain's completely pulling out the stops and I would feel remiss to not write it.

What if we dropped the idea of an intrinsic platform-specific pointer size and memory size altogether, and allowed platforms to define their memory spaces?

Something like this.
```zig
const avr = @import("avr");

const avr_code_pointer_size = @PointerSize(avr.code_memoryspace);
const avr_code_memory_size = @MemorySize(avr.code_memoryspace);
const avr_data_pointer_size = @PointerSize(avr.data_memoryspace);
const avr_data_memory_size = @MemorySize(avr.data_memoryspace);
```

User-defined names are obviously a no-go for cross-platform code, but any code that uses multiple memoryspaces wouldn't be cross-platform anyway, and at worst, converting a snippet to a single memoryspace architecture would be a find-replace.

**Comment by MasterQ32** [comment]
> Real Life Use Case? :thinking: 

There is sometimes the need to do pointer/type erasure. Primary use case would be anything OS-relevant (as in you're coding an OS). That's where you work a lot with pointers-as-numbers instead of actual memory slices.

> Is there any defense for isize and ipointer?

Yes! Memory/pointer distances. You cannot express an object size delta with `usize` and you cannot express a pointer distance with `upointer`. You cannot store *this object is 15 byte smaller and 240 byte before `X`* with only sized types.

I like the idea of `comptime` inspectible memory spaces, but `upointer` should be able to store *all* pointers and `usize` should be able to store all sizes.

**Comment by floopfloopfloopfloopfloop** [comment]
My thought was to use the specialized `msize` and `psize` as storage using the minimum amount of bits for its targeted memory space. e.g. If one memory space could allocate 2^16-1 bytes and the other 2^24-1, using a 24 bit value for both would waste memory more often than not?

**Comment by EleanorNB** [comment]
I agree with this, and I'd like to make one addition: `ulen`/`ilen`, for a register/ALU-sized int. Also, perhaps the architecture cannot load a register in a single instruction, or maybe it can load multiple -- so we could also have `udata`/`idata`, for a datapath-sized int.

So the complete list of ints would be:
- `uX`/`iX`: explicit-length signed/unsigned
- `ulen`/`ilen`: ALU-sized signed/unsigned (can be manipulated in one instruction)
- `udata`/`idata`: Datapath-sized signed/unsigned (can be loaded/stored in one instruction)
- `uptr`: Pointer-sized unsigned
- `usize`: Maximum size of an object in memory

(I don't think it makes sense to have `iptr`/`isize` when we have the others.)

**Comment by SpexGuy** [comment]
Re: `ualu` et al, It's not difficult to put register sizes in a platform header if you are compiling for a target where that's important.  IMO the language shouldn't though.  We decided not to do the C thing of changing our integer size between targets because it causes robustness problems -- suddenly your calculations may overflow where they didn't before.  For this reason, a "register sized integer" would generally be unsafe to use outside of very platform-specific code.  usize (and maybe upointer) is an important exception to this rule, because the language itself needs ways to talk about pointer-sized integers and object sizes that are guaranteed not to overflow (for these purposes) and are usable at runtime.  This is important for correctness.  The language does not need to talk about register sized objects though, because registers are not part of the zig abstract machine.  What's the size of a ualu or udata if compiling to JVM bytecode, for example?

In practice, for some targets, it will be important for performance to use register-sized integers.  But there are plenty of tools for tracking down performance problems, and Zig makes it easy to modify code in those locations to use faster integers.  Tracking down a rare integer overflow that only happens in production on one platform is far more difficult.

There might still be an argument for `ucell`, because the language does have the concept of addressable units.  But that seems like it belongs in a separate proposal.

**Comment by EleanorNB** [comment]
> There might still be an argument for ucell, because the language does have the concept of addressable units. But that seems like it belongs in a separate proposal.

As you command: #7693

**Comment by EleanorNB** [comment]
Specifically re. A64, those top bits aren't wasted -- they're used for pointer verification on newer versions. I think this is actually a plus: you could only get a pointer which is invalid in this way by casting a random integer, or going wayyy out of bounds, and the machine *should* yell at you if you do that.

Re. pointer comparison, if you want to compare pointers you can compare pointers -- IMO the compiler should be aware of memory segmentation and know how to check if pointers are *really* equal. No need to drag integers into it.

**Comment by kyle-github** [comment]
> Specifically re. A64, those top bits aren't wasted -- they're used for pointer verification on newer versions.

Oh, they are not wasted!   The original plan was to use those for [tagged pointers](https://en.wikichip.org/wiki/arm/aarch64#MMU) as is often done in Smalltalk, Lua etc.  More recently they are also used for [memory check tags](https://en.wikichip.org/wiki/memory_tagging) or whatever they are calling it these days.

By contrast AMD decided to go the other way and make it more painful to do anything with "unused" address bits.  I think the jury is still out about which one of these options was the better idea.

> Re. pointer comparison, if you want to compare pointers you can compare pointers -- IMO the compiler should be aware of memory segmentation and know how to check if pointers are really equal. No need to drag integers into it.

Think about walking a linked list looking for an element.   Every single time you move to a new node, you need to change your new pointer into normalized form for comparison.  Whether that is fabricating a 32-bit pointer from the segment and offset on a 16-bit x86 CPU or masking off bits on Arm CPUs, you'll need to do that.

As I just recently caught up to the discussion on Discord where @MasterQ32 gave a very good example of the AVR series and the "joys" of asymmetric Harvard architectures, I tried to expand on the ideas he had: I wrote up more in [a comment](https://github.com/ziglang/zig/issues/653#issuecomment-754899899) in issue #653.

**Comment by EleanorNB** [comment]
If we really don't want the top bits to be available, we can simply define `usize` to leave them out, i.e. it would be `u52` on A64 rather than `u64`.

Perhaps I wasn't clear: I think that pointer comparison should not be integer comparison of pointers; Zig has a stronger type system than C, we can distinguish pointers from integers. So, when comparing pointers, the compiler will know the platform and how to convert pointers into normal form or mask them off to compare them -- the user won't have to do that manually. This may mean that equality comparison becomes multiple machine instructions, but we're not aiming to be a macro assembler.

**Comment by ryanschneider** [comment]
> Is there any defense for `isize` and `ipointer`? The only time I can imagine a negative value in those contexts is to check for overflow, or something esoteric. The former can be done with existing operators. The latter could be done with an explicitly safe cast.

C functions that return pointers often use "negative pointers" for error conditions, for example `signal(2)` returns `SIG_ERR (-1)` to signal an error.  Not sure if this justifies the inclusion by itself but figured it was worth mentioning.

**Comment by oldwo** [comment]
If Zig aims to support diverse platforms, including some yet-unknown future platforms, no false promise should be made that all pointers are equal and can be converted back-and-forth to an integer. usize is exactly such a promise. I strongly favor the idea by @floopfloopfloopfloopfloop that platform headers should provide some function like
`const my_usize=@PointerSize(pointertype)`
And then you use that `my_usize` where you need it. There might also be functions to convert a pointer to an integer and back, but not all platforms are required to implement it.

**Comment by ethindp** [comment]
Slicing off the bits that we can ignore (e.g. bits 48:64 on x64) is a bad idea. Primarily because the "size" of a memory address can change at runtime. For example, if you've enabled paging but haven't set `CR4.LA57`, your address size is 48 bits (bits 48:64 are either zero or one). But if you set `CR4.LA57`, suddenly your address size changes to 57 bits (if I remember right) (bits 57:64 are zero or one). Or, even worse, the upper bits of an address can have meaning! For example, again on x86_64, if you enable protection keys, suddenly the upper 4 bits or so are the protection key for either user or supervisor mode. I don't really see how a strong type guarantee for `usize`/`isize` could be possible given that all of this can change just by flipping a few bits in a control register, and x64 isn't the only architecture that does this (RISC-V and ARM/Aarch64 do this, I'm pretty sure). You can't add probing of this into the language as a runtime feature, either, since, at least on x86 or x64, attempting to access a control register outside supervisor mode is illegal.

**Comment by iraniangethub** [comment]
You don't have permission to access "http://creativecommons.us4.list-manage.com/subscribe/post?" on this server.
Reference #18.1c284317.1697075770.5475041BEGIN:VCARD
VERSION;TYPE=WORK:3.0
FN;CHARSET=UTF-8:Creative Commons
N;CHARSET=UTF-8:Creative Commons;;;;
PROFILE:VCARD
ADR;CHARSET=UTF-8:;;211 Hope St # 1866;Mountain View;CA;94042-8817;USA
EMAIL:info@creativecommons.org
ORG;CHARSET=UTF-8:Creative Commons
URL:https://creativecommons.org
END:VCARD
CC General List
Email Sent
For security, we've sent an email to your inbox that contains a link to update your preferences.
Creative Commons
211 Hope St # 1866
Mountain View, CA 94042-8817

Add us to your address book
140 results for Re info support money dollar $ cash rule Xchanged Rial Iranian Amount 10,990,000,443 Rial Iranian documents BIC: INGBNL2A IBAN: NL82INGB0006484837send money cash rule delivery accept money from releases transaction confirmation from SMS phone number 09128184280 account number bank mellat Iranian IR09 0120 0200 0000 9333 8020 15 name nosrat rafiepoor card number 6104337322967389 status include... IP Statik 193.174.13.255CC General List
Email Sent
For security, we've sent an email to your inbox that contains a link to update your preferences.
Creative Commons
211 Hope St # 1866
Mountain View, CA 94042-8817

Add us to CC General List
Email Sent
For security, we've sent an email to your inbox that contains a link to update your preferences.
Creative Commons
211 Hope St # 1866
Mountain View, CA 94042-8817

Add us to your address book address book

**Comment by perillo** [comment]
The C language also has the `ptrdiff_t` type, the type of the difference between two pointers.

See `https://stackoverflow.com/questions/1464174/size-t-vs-uintptr-t for more details, especially the answer by Alex Martelli https://stackoverflow.com/a/1464194.

If #1738 is approved, `ptrdiff` should be used.

**Comment by Timmmm** [comment]
I dunno if you guys have also considered CHERI. If you aren't familiar with it, the TL;DR is that pointers are twice as big - the bottom half is the same (stores the address). The top half stores compressed [bottom, top) bounds (and some other stuff). There's an out-of-band tag bit to indicate that pointers are valid, so you can't just change the bounds by fiat.

The awkward consequences are:

🔧 addresses and pointers are different sizes.
🔧 pointer provenance is not just some high level optimisation thing any more; you actually *can't* cast a pointer to an integer and back, even in assembly. Nor can you create a pointer to one object by adding to a pointer to a different object.
🔧 pointers can have the same address, but still not be equal.

Rust has [done some work on this recently](https://blog.rust-lang.org/2025/01/09/Rust-1.84.0.html#strict-provenance-apis). Maybe it could be... borrowed. :-D

**Comment by ikskuh** [comment]
> pointer provenance is not just some high level optimisation thing any more; you actually _can't_ cast a pointer to an integer and back, even in assembly. Nor can you create a pointer to one object by adding to a pointer to a different object.

could you explain how i would create a pointer from an integer? Because it's a typical use case for a lot of lowlevel tasks. feel free to answer on discord or zulip, as it's kinda offtopic here

**Comment by Timmmm** [comment]
> could you explain how i would create a pointer from an integer? Because it's a typical use case for a lot of lowlevel tasks. feel free to answer on discord or zulip, as it's kinda offtopic here

You can't, so it is true that running some code on CHERI requires a bit of code modification (to 0.026% of the lines of code according to previous ports).

What you have to do instead is retain a capability (pointer) that has bounds that cover all of the addresses you might want to use (e.g. some arena probably, or the entire heap if you want) and then you can derive a pointer from that "parent" capability to anywhere within it using [`scaddr`](https://riscv.github.io/riscv-cheri/#SCADDR) or [`cadd`](https://riscv.github.io/riscv-cheri/#CADD) (i.e. you can add to a pointer and it will still work as long as the pointer is within bounds).

It is possible to have a "god capability" with unlimited bounds that you can use to turn any integer into a valid point, but that kind of defeats the point!

(Sorry for not using Zulip - on my phone.)
